crates/wasi-common/src/old/snapshot_0/sys/windows/hostcalls_impl/fs.rs

```diff
@@ -388,44 +387,46 @@ pub(crate) fn path_rename(resolved_old: PathGet, resolved_new: PathGet) -> Resul
     }
 
     // TODO handle symlinks
-
-    fs::rename(&old_path, &new_path).or_else(|e| match e.raw_os_error() {
-        Some(e) => {
-            use winx::winerror::WinError;
-
-            log::debug!("path_rename at rename error code={:?}", e);
-            match WinError::from_u32(e as u32) {
-                WinError::ERROR_ACCESS_DENIED => {
+    let err = match fs::rename(&old_path, &new_path) {
+        Ok(()) => return Ok(()),
+        Err(e) => e,
+    };
+    match err.raw_os_error() {
+        Some(code) => {
+            log::debug!("path_rename at rename error code={:?}", code);
+            match code as u32 {
+                winerror::ERROR_ACCESS_DENIED => {
                     // So most likely dealing with new_path == dir.
                     // Eliminate case old_path == file first.
                     if old_path.is_file() {
-                        Err(Error::EISDIR)
+                        return Err(Error::EISDIR);
                     } else {
                         // Ok, let's try removing an empty dir at new_path if it exists
                         // and is a nonempty dir.
-                        fs::remove_dir(&new_path)
-                            .and_then(|()| fs::rename(old_path, new_path))
-                            .map_err(Into::into)
+                        fs::remove_dir(&new_path)?;
+                        fs::rename(old_path, new_path)?;
+                        return Ok(());
                     }
                 }
-                WinError::ERROR_INVALID_NAME => {
+                winerror::ERROR_INVALID_NAME => {
                     // If source contains trailing slashes, check if we are dealing with
                     // a file instead of a dir, and if so, throw ENOTDIR.
                     if let Some(path) = strip_trailing_slashes_and_concatenate(&resolved_old)? {
                         if path.is_file() {
                             return Err(Error::ENOTDIR);
                         }
                     }
-                    Err(WinError::ERROR_INVALID_NAME.into())
                 }
-                e => Err(e.into()),
+                _ => {}
```

Comment:
I'd say that without `return`s the control flow is easier to follow.

Reply:
See above :-)